### PR TITLE
PUBDEV-5265 - GLM crashes on an empty categorial variable

### PR DESCRIPTION
**Pull request by Pscheidl**

### CODE REVIEW

**Comment by Pscheidl** [comment]
If you merge it into your branch @wendycwong , it will fix the test no6 in pyunit_glm_multinomial_large.py

**Comment by michalkurka** [comment]
@Pscheidl I reverted locally your change in DataInfo, built H2O and ran with assertions enabled. The test passed.

Is this 100% reproducible? I am running on Python 2.7

**Comment by Pscheidl** [comment]
@michalkurka I retried my test and got the same results as you did. However, the original test pyunit_glm_multinomial_large did fail. Seems like the fix is valid but the test is invalid. I'll address that.

File: h2o-algos/src/main/java/hex/DataInfo.java [hunk 1]
```diff
@@ -973,7 +973,6 @@ public final int getCategoricalId(int cid, int val) {
     int [] offs = fullCatOffsets();
     int expandedVal = val + offs[cid];
     if(expandedVal >= offs[cid+1]) {  // previously unseen level
-      assert _valid:"Categorical value out of bounds, got " + val + ", next cat starts at " + fullCatOffsets()[cid+1];
```
Comment (wendycwong) [inline h2o-algos/src/main/java/hex/DataInfo.java]: Looks like this address the issue with the validation set containing enum values that are not found in training data set.  It just throw an error.  Just talked to Megan K, usually, for test dataset, it will replace the unseen enum value with the mode.  By commenting out this assertion, it should be okay.

**Comment by wendycwong** [comment]
Pavel:

Yes, your fix should fix this problem.  It is just strange nobody has caught this before.  


Thanks,
Wendy

File: h2o-algos/src/main/java/hex/DataInfo.java [hunk 2]
```diff
@@ -967,13 +967,16 @@ public final int getCategoricalId(int cid, int val) {
...
+    }
     if(val < 0) return -1; // column si to be skipped
     int [] offs = fullCatOffsets();
     int expandedVal = val + offs[cid];
     if(expandedVal >= offs[cid+1]) {  // previously unseen level
-      assert _valid:"Categorical value out of bounds, got " + val + ", next cat starts at " + fullCatOffsets()[cid+1];
+      assert _valid : "Categorical value out of bounds, got " + val + ", next cat starts at " + fullCatOffsets()[cid + 1];
       if(_skipMissing)
         return -1;
       val = _catNAFill[cid];
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/DataInfo.java]: This is also not correct. Anytime you use _catNAFill, you need to consider the adjustment of indices.

    if (!_useAllFactorLevels && !isIWV) {  // categorical interaction vecs drop reference level in a special way
      val = val - 1;
    }

Comment (Pscheidl) [reply h2o-algos/src/main/java/hex/DataInfo.java]: This one has been fixed.

Comment (michalkurka) [reply h2o-algos/src/main/java/hex/DataInfo.java]: ARE YOU SURE? :P

**Comment by michalkurka** [comment]
@Pscheidl please rebase the PR

I think it looks good, lets get green tests now

File: h2o-algos/src/main/java/hex/DataInfo.java (hunk 2 above)
Comment (Pscheidl) [reply h2o-algos/src/main/java/hex/DataInfo.java]: Yes ! 💃

**Comment by Pscheidl** [comment]
The test introduced in this PR is broken on Python 3

**Comment by Pscheidl** [comment]
Rebased onto master for latest fixes.

**Comment by wendycwong** [comment]
Looks good to me.
